# 1. 生成categories字段

## 先合并所有csv

In [1]:
import pandas as pd
import os

# 定义数据集的根目录
root_dir = r"D:\桌面\Kuzushiji\数据集\allin"

# 创建一个空的DataFrame来存储合并后的数据
merged_df = pd.DataFrame()

# 遍历根目录下的所有文件夹，每个文件夹对应一本kuzushiji书
for folder in os.listdir(root_dir):
    # 获取当前文件夹的路径
    folder_path = os.path.join(root_dir, folder)
    # 获取当前文件夹下的csv文件的路径
    csv_path = os.path.join(folder_path, folder + "_coordinate.csv")
    # 读取csv文件为一个DataFrame
    csv_df = pd.read_csv(csv_path)
    # 删去Block ID和Char ID两列
    csv_df.drop(columns=["Block ID", "Char ID"], inplace=True)
    # 将当前DataFrame追加到合并后的DataFrame中
    merged_df = pd.concat([merged_df, csv_df], ignore_index=True) # 修改这一行

# 将合并后的DataFrame保存为一个新的csv文件
merged_df.to_csv("merged_coordinate.csv", index=False)


FileNotFoundError: [WinError 3] 系统找不到指定的路径。: 'D:\\桌面\\Kuzushiji\\数据集\\allin'

In [2]:
import pandas as pd

# 读取合并后的csv文件为一个DataFrame
merged_df = pd.read_csv("merged_coordinate.csv")

# 提取第一列的Unicode为一个Series
unicode_series = merged_df["Unicode"]

# 删除重复的值，并转换为一个列表
unicode_list = unicode_series.drop_duplicates().tolist()

# 计算列表的长度
unicode_length = len(unicode_list)

# 打印结果
print("Unicode列表的长度为：", unicode_length)


Unicode列表的长度为： 4328


谢谢。你能帮我把这个unicode_list转化为coco数据集格式中的categories字段吗？只需用id和name，id从0开始。要求顺序是按照Unicode值从小到大排列


In [3]:
# 对列表按照Unicode值从小到大排序
unicode_list.sort()
unicode_list

['U+0031',
 'U+0032',
 'U+0034',
 'U+0036',
 'U+0039',
 'U+003F',
 'U+2000B',
 'U+20D45',
 'U+2123D',
 'U+22999',
 'U+22CE3',
 'U+231C3',
 'U+2324A',
 'U+233D0',
 'U+23D0E',
 'U+241C6',
 'U+242CE',
 'U+2437D',
 'U+24E30',
 'U+24E60',
 'U+24FA3',
 'U+2564A',
 'U+25750',
 'U+25877',
 'U+25A0',
 'U+25A1',
 'U+25B2',
 'U+25B3',
 'U+25CB',
 'U+25CF',
 'U+25DA1',
 'U+25E2E',
 'U+25EF',
 'U+25FFB',
 'U+264B3',
 'U+266B0',
 'U+27752',
 'U+28263',
 'U+29780',
 'U+29DDA',
 'U+29E75',
 'U+3002',
 'U+3005',
 'U+3006',
 'U+3007',
 'U+300C',
 'U+300D',
 'U+3031',
 'U+3032',
 'U+303B',
 'U+303D',
 'U+3042',
 'U+3044',
 'U+3046',
 'U+3048',
 'U+304A',
 'U+304B',
 'U+304C',
 'U+304D',
 'U+304E',
 'U+304F',
 'U+3050',
 'U+3051',
 'U+3052',
 'U+3053',
 'U+3054',
 'U+3055',
 'U+3056',
 'U+3057',
 'U+3058',
 'U+3059',
 'U+305A',
 'U+305B',
 'U+305C',
 'U+305D',
 'U+305E',
 'U+305F',
 'U+3060',
 'U+3061',
 'U+3062',
 'U+3063',
 'U+3064',
 'U+3065',
 'U+3066',
 'U+3067',
 'U+3068',
 'U+3069',
 'U+306A',
 'U+

In [4]:
# 保存该list
with open("Unicode_new.txt","w") as f:
    f.write(",".join("'" + item + "'" for item in unicode_list))

In [5]:
# 创建一个空的列表来存储categories字段
nitian = []

# 遍历unicode_list，为每个Unicode生成一个字典，包含id和name
for i, u in enumerate(unicode_list):
    # id从0开始，name是Unicode对应的汉字
    n = {"id": i, "name": chr(int(u[2:], 16))}
    # 将字典追加到categories列表中
    nitian.append(n)
nitian

[{'id': 0, 'name': '1'},
 {'id': 1, 'name': '2'},
 {'id': 2, 'name': '4'},
 {'id': 3, 'name': '6'},
 {'id': 4, 'name': '9'},
 {'id': 5, 'name': '?'},
 {'id': 6, 'name': '𠀋'},
 {'id': 7, 'name': '𠵅'},
 {'id': 8, 'name': '𡈽'},
 {'id': 9, 'name': '𢦙'},
 {'id': 10, 'name': '𢳣'},
 {'id': 11, 'name': '𣇃'},
 {'id': 12, 'name': '𣉊'},
 {'id': 13, 'name': '𣏐'},
 {'id': 14, 'name': '𣴎'},
 {'id': 15, 'name': '𤇆'},
 {'id': 16, 'name': '𤋎'},
 {'id': 17, 'name': '𤍽'},
 {'id': 18, 'name': '𤸰'},
 {'id': 19, 'name': '𤹠'},
 {'id': 20, 'name': '𤾣'},
 {'id': 21, 'name': '𥙊'},
 {'id': 22, 'name': '𥝐'},
 {'id': 23, 'name': '𥡷'},
 {'id': 24, 'name': '■'},
 {'id': 25, 'name': '□'},
 {'id': 26, 'name': '▲'},
 {'id': 27, 'name': '△'},
 {'id': 28, 'name': '○'},
 {'id': 29, 'name': '●'},
 {'id': 30, 'name': '𥶡'},
 {'id': 31, 'name': '𥸮'},
 {'id': 32, 'name': '◯'},
 {'id': 33, 'name': '𥿻'},
 {'id': 34, 'name': '𦒳'},
 {'id': 35, 'name': '𦚰'},
 {'id': 36, 'name': '𧝒'},
 {'id': 37, 'name': '𨉣'},
 {'id': 38, 'name': '𩞀

In [6]:
# 创建一个空的列表来存储categories字段
categories = []

# 遍历unicode_list，为每个Unicode生成一个字典，包含id和name
for i, u in enumerate(unicode_list):
    # id从0开始，name是Unicode对应的汉字
    category = {"id": i, "name": u}
    # 将字典追加到categories列表中
    categories.append(category)
categories

[{'id': 0, 'name': 'U+0031'},
 {'id': 1, 'name': 'U+0032'},
 {'id': 2, 'name': 'U+0034'},
 {'id': 3, 'name': 'U+0036'},
 {'id': 4, 'name': 'U+0039'},
 {'id': 5, 'name': 'U+003F'},
 {'id': 6, 'name': 'U+2000B'},
 {'id': 7, 'name': 'U+20D45'},
 {'id': 8, 'name': 'U+2123D'},
 {'id': 9, 'name': 'U+22999'},
 {'id': 10, 'name': 'U+22CE3'},
 {'id': 11, 'name': 'U+231C3'},
 {'id': 12, 'name': 'U+2324A'},
 {'id': 13, 'name': 'U+233D0'},
 {'id': 14, 'name': 'U+23D0E'},
 {'id': 15, 'name': 'U+241C6'},
 {'id': 16, 'name': 'U+242CE'},
 {'id': 17, 'name': 'U+2437D'},
 {'id': 18, 'name': 'U+24E30'},
 {'id': 19, 'name': 'U+24E60'},
 {'id': 20, 'name': 'U+24FA3'},
 {'id': 21, 'name': 'U+2564A'},
 {'id': 22, 'name': 'U+25750'},
 {'id': 23, 'name': 'U+25877'},
 {'id': 24, 'name': 'U+25A0'},
 {'id': 25, 'name': 'U+25A1'},
 {'id': 26, 'name': 'U+25B2'},
 {'id': 27, 'name': 'U+25B3'},
 {'id': 28, 'name': 'U+25CB'},
 {'id': 29, 'name': 'U+25CF'},
 {'id': 30, 'name': 'U+25DA1'},
 {'id': 31, 'name': 'U+25E2E'}

In [7]:
len(categories)

4328

# 2. 删去没有标注的照片

你好bing，我正在处理kuzushiji数据集来进行目标检测任务。我的数据集文件结构比较复杂，它是由若干个文件夹组成，每个文件夹的名是一本kuzushiji书，里面又包含一个images文件（里面是训练集照片）和一个csv文件（表示该本书上所有标注的坐标框等信息），格式如下：
D:\桌面\Kuzushiji\数据集\allin
├── 100241706
│   ├── images
│   │   ├── 100241706_00001_1.jpg
│   │   ├── 100241706_00002_1.jpg
│   │   ├── 100241706_00003_1.jpg
│   │   ├── 100241706_00004_2.jpg
│   │   └── ...
│   └── 100241706_coordinate.csv
├── 100249371
│   ├── images
│   │   ├── 100249371_00001_1.jpg
│   │   ├── 100249371_00002_1.jpg
│   │   ├── 100249371_00003_1.jpg
│   │   ├── 100249371_00004_2.jpg
│   │   └── ...
│   └── 100249371_coordinate.csv
└── ...

我有一些照片，这些照片ground truth是没有任何标注的，我想把他们筛选出去。

每个csv文件的格式是这样的：
Unicode	Image	X	Y	Block ID	Char ID	Width	Height
U+81EA	100241706_00004_2	1852	1736	B0001	C0001	104	219
U+5E8F	100241706_00004_2	1816	2096	B0001	C0002	152	296
U+82E5	100241706_00004_2	1465	951	B0001	C0003	172	216
U+3044	100241706_00004_2	1495	1218	B0001	C0004	143	69
U+6642	100241706_00004_2	1465	1338	B0001	C0005	168	179
csv文件的文件名是：书名_coordinate.csv, 例如D:\桌面\Kuzushiji\数据集\all\100241706\100241706_coordinate.csv
你能帮我把所有这些csv文件合并起来吗（重复的表头删去）

### 首先提取所有照片的名字（注意包括训练集+验证集）


In [8]:
import pandas as pd

# 读取合并后的csv文件为一个DataFrame
merged_df = pd.read_csv("merged_coordinate.csv")

# 提取Image列为一个Series
image_series = merged_df["Image"]

# 删除重复的值，并转换为一个列表
image_list = image_series.drop_duplicates().tolist()

In [9]:
# 打印结果
print("照片名列表为：", image_list)

照片名列表为： ['100241706_00004_2', '100241706_00005_1', '100241706_00005_2', '100241706_00006_1', '100241706_00007_2', '100241706_00008_1', '100241706_00008_2', '100241706_00009_1', '100241706_00009_2', '100241706_00010_1', '100241706_00010_2', '100241706_00011_1', '100241706_00011_2', '100241706_00012_1', '100241706_00012_2', '100241706_00013_1', '100241706_00013_2', '100241706_00014_1', '100241706_00014_2', '100241706_00015_1', '100241706_00015_2', '100241706_00016_1', '100241706_00016_2', '100241706_00017_1', '100241706_00017_2', '100241706_00018_1', '100241706_00018_2', '100241706_00019_1', '100241706_00019_2', '100241706_00020_1', '100241706_00020_2', '100241706_00021_1', '100241706_00021_2', '100241706_00022_1', '100241706_00022_2', '100241706_00023_1', '100241706_00023_2', '100241706_00024_1', '100241706_00024_2', '100241706_00025_1', '100241706_00025_2', '100241706_00026_1', '100241706_00026_2', '100241706_00027_1', '100241706_00027_2', '100241706_00028_1', '100241706_00028_2', '100

我发现有一些训练集照片的ground truth是没有任何标注的，也就是在标注的csv文件当中没有任何标注。我想把他们筛选出去，放到一个Nolabel的文件夹下，image_list 你能帮我生成一份代码吗

我可以帮你把all_images下所有的照片文件名提取出来作为一个列表。你可以使用以下的python代码来实现这个功能：

In [10]:
import pandas as pd
import os
import shutil

# 读取合并后的csv文件为一个DataFrame
merged_df = pd.read_csv("merged_coordinate.csv")

# 提取Image列为一个Series
image_series = merged_df["Image"]

# 删除重复的值，并转换为一个列表
image_list = image_series.drop_duplicates().tolist()

# 定义数据集的根目录
root_dir = r"D:\桌面\Kuzushiji\数据集\allin"

# 定义Nolabel的文件夹路径
nolabel_dir = "Nolabel"

# 如果Nolabel的文件夹不存在，就创建一个
if not os.path.exists(nolabel_dir):
    os.mkdir(nolabel_dir)

# 遍历根目录下的所有文件夹，每个文件夹对应一本kuzushiji书
for folder in os.listdir(root_dir):
    # 获取当前文件夹的路径
    folder_path = os.path.join(root_dir, folder)
    # 获取当前文件夹下的images文件夹的路径
    images_path = os.path.join(folder_path, "images")
    # 遍历images文件夹下的所有图片文件
    for image in os.listdir(images_path):
        # 获取当前图片文件的路径
        image_path = os.path.join(images_path, image)
        # 判断当前图片是否在image_list里
        if image[:-4] not in image_list: # 修改这一行，去掉jpg后缀再判断
            print("Move")
            # 如果不在，就将图片移动到Nolabel的文件夹下
            shutil.move(image_path, nolabel_dir)

FileNotFoundError: [WinError 3] 系统找不到指定的路径。: 'D:\\桌面\\Kuzushiji\\数据集\\allin'

### 筛选完毕
15本测试集，1本验证集（源氏物语），28本训练集。

首先把所有test照片提取出来

In [11]:
import os
import shutil

# 定义数据集的根目录
root_dir = "test_origin"

# 定义目标文件夹的路径
target_dir = "test"

# 如果目标文件夹不存在，就创建一个
if not os.path.exists(target_dir):
    os.mkdir(target_dir)

# 遍历根目录下的所有文件夹，每个文件夹对应一本kuzushiji书
for folder in os.listdir(root_dir):
    # 获取当前文件夹的路径
    folder_path = os.path.join(root_dir, folder)
    # 获取当前文件夹下的images文件夹的路径
    images_path = os.path.join(folder_path, "images")
    # 遍历images文件夹下的所有图片文件
    for image in os.listdir(images_path):
        # 获取当前图片文件的路径
        image_path = os.path.join(images_path, image)
        # 将图片复制到目标文件夹下
        shutil.copy(image_path, target_dir)


### 对比一下两个新老版本的测试集

In [12]:
import os

# 定义第一个文件夹的路径
file1_dir = r"D:\mmdetection\kuzuTest\clean_data\test"

# 定义第二个文件夹的路径
file2_dir = r"D:\mmdetection\kuzuTest\NewCOCO\test" 

# 创建一个空的列表来存储第一个文件夹里的照片名
file1_list = []

# 遍历第一个文件夹下的所有图片文件
for image in os.listdir(file1_dir):
    # 将图片文件名追加到列表中
    file1_list.append(image)

# 创建一个空的列表来存储第二个文件夹里的照片名
file2_list = []

# 遍历第二个文件夹下的所有图片文件
for image in os.listdir(file2_dir):
    # 将图片文件名追加到列表中
    file2_list.append(image)

# 使用set函数，找出两个列表中共同存在的照片名，即交集
common_list = list(set(file1_list) & set(file2_list))

# 使用set函数，找出两个列表中不同的照片名，即对称差集
different_list = list(set(file1_list) ^ set(file2_list))

# 打印结果
print("两个文件里共同存在的照片名有：", common_list)
print("两个文件里不同的照片名有：", different_list)


两个文件里共同存在的照片名有： ['200008316_00078_1.jpg', '200019865_00010_1.jpg', '200019865_00023_1.jpg', '200005798_00046_2.jpg', '200021071_00083_1.jpg', '200019865_00035_2.jpg', '200008316_00113_2.jpg', '200021086_00015_1.jpg', '200021071_00076_2.jpg', '200021071_00104_2.jpg', '200021063_00027_1.jpg', '200021071_00044_1.jpg', '200020019_00004_1.jpg', '200008316_00083_1.jpg', '200017458_00019_2.jpg', '200005798_00033_1.jpg', '200020019_00005_2.jpg', '200021086_00010_1.jpg', '200005798_00103_1.jpg', '200018243_00011_1.jpg', '200008316_00074_1.jpg', '200006665_00007_2.jpg', '200008316_00073_2.jpg', '200008316_00052_2.jpg', '200025191_00005_2.jpg', '200008316_00075_2.jpg', '200008316_00011_2.jpg', '200005798_00057_2.jpg', '200005798_00101_1.jpg', '200015843_00115_2.jpg', '200021071_00058_2.jpg', '200008316_00060_2.jpg', '200008316_00041_2.jpg', '200020019_00032_2.jpg', '200021071_00009_2.jpg', '200021071_00099_2.jpg', '200018243_00004_2.jpg', '200005798_00060_2.jpg', '200005798_00089_1.jpg', '2000066

# 3. 测试集处理

## 首先依然是合并所有测试集csv

In [13]:
import pandas as pd
import os

# 定义数据集的根目录
root_dir = r"test_origin"

# 创建一个空的DataFrame来存储合并后的数据
merged_test_df = pd.DataFrame()

# 遍历根目录下的所有文件夹，每个文件夹对应一本kuzushiji书
for folder in os.listdir(root_dir):
    # 获取当前文件夹的路径
    folder_path = os.path.join(root_dir, folder)
    # 获取当前文件夹下的csv文件的路径
    csv_path = os.path.join(folder_path, folder + "_coordinate.csv")
    # 读取csv文件为一个DataFrame
    csv_df = pd.read_csv(csv_path)
    # 删去Block ID和Char ID两列
    csv_df.drop(columns=["Block ID", "Char ID"], inplace=True)
    # 将当前DataFrame追加到合并后的DataFrame中
    merged_test_df = pd.concat([merged_test_df, csv_df], ignore_index=True) # 修改这一行

# 将合并后的DataFrame保存为一个新的csv文件
merged_test_df.to_csv("merged_test_coordinate.csv", index=False)


In [14]:
merged_test_df

Unicode              Image     X     Y  Width  Height
0       U+4E8C  200004107_00003_2  1514  1081    122      85
1       U+4EBA  200004107_00003_2  1513  1186    139     178
2       U+6BD4  200004107_00003_2  1524  1371    134     115
3       U+4E18  200004107_00003_2  1510  1519    136     117
4       U+5C3C  200004107_00003_2  1520  1697    139     117
...        ...                ...   ...   ...    ...     ...
391041  U+306B  200025191_00069_2   915  1828     72     132
391042  U+304F  200025191_00069_2   913  1966     71     104
391043  U+308C  200025191_00069_2   888  2069    112     164
391044  U+3057  200025191_00069_2   950  2239     38     165
391045  U+3092  200025191_00069_2   886  2373     93     155

[391046 rows x 6 columns]

In [15]:
import pandas as pd
import os
from PIL import Image


# 提取Image列为一个Series
test_series = merged_test_df["Image"]

# 删除重复的值，并转换为一个列表
test_list = test_series.drop_duplicates().tolist()

# 定义test_images的文件夹路径
test_images_dir = "test"

test_list

['200004107_00003_2',
 '200004107_00004_1',
 '200004107_00005_1',
 '200004107_00006_1',
 '200004107_00006_2',
 '200004107_00007_1',
 '200004107_00008_1',
 '200004107_00008_2',
 '200004107_00009_1',
 '200004107_00010_1',
 '200004107_00010_2',
 '200004107_00012_1',
 '200004107_00012_2',
 '200004107_00013_2',
 '200004107_00014_2',
 '200004107_00015_1',
 '200004107_00016_1',
 '200004107_00016_2',
 '200004107_00017_1',
 '200004107_00017_2',
 '200004107_00018_2',
 '200004107_00019_1',
 '200004107_00019_2',
 '200004107_00020_1',
 '200004107_00020_2',
 '200004107_00021_1',
 '200004107_00022_1',
 '200004107_00023_1',
 '200004107_00024_1',
 '200004107_00025_1',
 '200004107_00026_1',
 '200004107_00026_2',
 '200004107_00027_2',
 '200004107_00028_2',
 '200004107_00029_2',
 '200004107_00030_1',
 '200004107_00031_1',
 '200004107_00032_1',
 '200004107_00032_2',
 '200004107_00033_2',
 '200004107_00034_2',
 '200004107_00035_1',
 '200004107_00036_1',
 '200004107_00036_2',
 '200004107_00037_1',
 '20000410

In [16]:
# 创建一个空的列表来存储coco数据集格式的images字段
coco_test = []

# 遍历image_list中的每个照片名
for image in test_list:
    # 获取当前照片文件的路径
    image_path = os.path.join(test_images_dir, image+'.jpg')
    # 打开当前照片文件，获取宽度和高度
    img = Image.open(image_path)
    width, height = img.size
    # 关闭当前照片文件
    img.close()
    # 创建一个字典，存储当前照片的信息，包括id，file_name，width和height
    image_info = {
        "file_name": image+'.jpg',
        "id": test_list.index(image), # id从0开始递增
        "height": height,
        "width": width
    }
    # 将字典追加到coco_images列表中
    coco_test.append(image_info)

# 打印结果
print("coco数据集格式的images字段为：", coco_test)

coco数据集格式的images字段为： [{'file_name': '200004107_00003_2.jpg', 'id': 0, 'height': 2962, 'width': 2026}, {'file_name': '200004107_00004_1.jpg', 'id': 1, 'height': 2934, 'width': 1998}, {'file_name': '200004107_00005_1.jpg', 'id': 2, 'height': 2946, 'width': 2014}, {'file_name': '200004107_00006_1.jpg', 'id': 3, 'height': 2975, 'width': 1990}, {'file_name': '200004107_00006_2.jpg', 'id': 4, 'height': 2983, 'width': 2050}, {'file_name': '200004107_00007_1.jpg', 'id': 5, 'height': 2951, 'width': 2006}, {'file_name': '200004107_00008_1.jpg', 'id': 6, 'height': 2950, 'width': 2018}, {'file_name': '200004107_00008_2.jpg', 'id': 7, 'height': 2945, 'width': 2002}, {'file_name': '200004107_00009_1.jpg', 'id': 8, 'height': 2955, 'width': 2026}, {'file_name': '200004107_00010_1.jpg', 'id': 9, 'height': 2938, 'width': 1986}, {'file_name': '200004107_00010_2.jpg', 'id': 10, 'height': 2942, 'width': 1994}, {'file_name': '200004107_00012_1.jpg', 'id': 11, 'height': 2978, 'width': 2058}, {'file_name': '2

In [17]:
# 创建一个空的字典来存储images字段中的照片名和id的对应关系
image_dict = {}

# 遍历images字段中的每个字典
for image_info in coco_test:
    # 获取当前字典中的照片名和id
    image_name = image_info["file_name"]
    image_id = image_info["id"]
    # 将照片名和id作为键值对存入image_dict中
    image_dict[image_name] = image_id

# 创建一个空的字典来存储categories字段中的类别名和id的对应关系
category_dict = {}

# 遍历categories字段中的每个字典
for category_info in categories:
    # 获取当前字典中的类别名和id
    category_name = category_info["name"]
    category_id = category_info["id"]
    # 将类别名和id作为键值对存入category_dict中
    category_dict[category_name] = category_id


In [ ]:
category_dict

In [ ]:
image_dict

In [18]:
# 创建一个新的空字典
new_image_dict = {}

# 遍历原字典的键值对
for key, value in image_dict.items():
    # 去掉键中的.jpg
    new_key = key.replace(".jpg", "")
    # 把新的键值对添加到新字典中
    new_image_dict[new_key] = value
new_image_dict 

{'200004107_00003_2': 0,
 '200004107_00004_1': 1,
 '200004107_00005_1': 2,
 '200004107_00006_1': 3,
 '200004107_00006_2': 4,
 '200004107_00007_1': 5,
 '200004107_00008_1': 6,
 '200004107_00008_2': 7,
 '200004107_00009_1': 8,
 '200004107_00010_1': 9,
 '200004107_00010_2': 10,
 '200004107_00012_1': 11,
 '200004107_00012_2': 12,
 '200004107_00013_2': 13,
 '200004107_00014_2': 14,
 '200004107_00015_1': 15,
 '200004107_00016_1': 16,
 '200004107_00016_2': 17,
 '200004107_00017_1': 18,
 '200004107_00017_2': 19,
 '200004107_00018_2': 20,
 '200004107_00019_1': 21,
 '200004107_00019_2': 22,
 '200004107_00020_1': 23,
 '200004107_00020_2': 24,
 '200004107_00021_1': 25,
 '200004107_00022_1': 26,
 '200004107_00023_1': 27,
 '200004107_00024_1': 28,
 '200004107_00025_1': 29,
 '200004107_00026_1': 30,
 '200004107_00026_2': 31,
 '200004107_00027_2': 32,
 '200004107_00028_2': 33,
 '200004107_00029_2': 34,
 '200004107_00030_1': 35,
 '200004107_00031_1': 36,
 '200004107_00032_1': 37,
 '200004107_00032_2': 

In [20]:
# 创建一个空的列表来存储coco数据集格式的annotations字段
coco_annotations = []

# 遍历merged_df中的每一行
for index, row in merged_test_df.iterrows():
    # 获取当前行的照片名，类别名，坐标和面积
    image_name = row["Image"]
    category_name = row["Unicode"]
    bbox = [row["X"], row["Y"], row["Width"], row["Height"]]
    area = row["Width"] * row["Height"]
    # 创建一个字典，存储当前行的信息，包括id，image_id，category_id，bbox和area
    
    annotation_info = {
        "id": index, # id从0开始递增
        "image_id": new_image_dict[image_name], # 根据照片名在new_image_dict中查找对应的id
        "category_id": category_dict[category_name], # 根据类别名在category_dict中查找对应的id
        "bbox": bbox,
        "area": area,
        "iscrowd": 0
    }
    # 将字典追加到coco_annotations列表中
    coco_annotations.append(annotation_info)

In [21]:
# 设置一个变量n，表示输出的数量
n = 10

# 打印结果的前n个元素
print("coco数据集格式的annotations字段的前{}个元素为：".format(n), coco_annotations[:n])


coco数据集格式的annotations字段的前10个元素为： [{'id': 0, 'image_id': 0, 'category_id': 293, 'bbox': [1514, 1081, 122, 85], 'area': 10370, 'iscrowd': 0}, {'id': 1, 'image_id': 0, 'category_id': 311, 'bbox': [1513, 1186, 139, 178], 'area': 24742, 'iscrowd': 0}, {'id': 2, 'image_id': 0, 'category_id': 1992, 'bbox': [1524, 1371, 134, 115], 'area': 15410, 'iscrowd': 0}, {'id': 3, 'image_id': 0, 'category_id': 259, 'bbox': [1510, 1519, 136, 117], 'area': 15912, 'iscrowd': 0}, {'id': 4, 'image_id': 0, 'category_id': 1094, 'bbox': [1520, 1697, 139, 117], 'area': 16263, 'iscrowd': 0}, {'id': 5, 'image_id': 0, 'category_id': 1757, 'bbox': [1332, 632, 120, 145], 'area': 17400, 'iscrowd': 0}, {'id': 6, 'image_id': 0, 'category_id': 83, 'bbox': [1337, 780, 91, 118], 'area': 10738, 'iscrowd': 0}, {'id': 7, 'image_id': 0, 'category_id': 1082, 'bbox': [1339, 908, 112, 128], 'area': 14336, 'iscrowd': 0}, {'id': 8, 'image_id': 0, 'category_id': 1679, 'bbox': [1350, 1008, 97, 108], 'area': 10476, 'iscrowd': 0}, {'id'

In [22]:
test_dict_to_save = {'annotations':coco_annotations, 'images':coco_test, 'categories':categories}

In [23]:
import json
# 日语字符以utf-8编码了，可以正常读取！
with open("test.json", 'w',encoding='utf-8') as f:
    json.dump(test_dict_to_save, f)

### 生成测试集kaggle格式Ground Truth

我们之前生成的那个应该也没问题，因为数据集完全一样，不过保险起见再重新做一次

你好bing，你能帮我把我自己的标注框merged_test_df（格式如下）
Unicode	Image	X	Y	Width	Height
0	U+4E8C	200004107_00003_2	1514	1081	122	85
1	U+4EBA	200004107_00003_2	1513	1186	139	178
2	U+6BD4	200004107_00003_2	1524	1371	134	115
3	U+4E18	200004107_00003_2	1510	1519	136	117
4	U+5C3C	200004107_00003_2	1520	1697	139	117
...	...	...	...	...	...	...
变成kaggle submit的格式吗：
image_id	labels
200004107_00003_2	U+4E8C 1514 1081 122 85 U+4EBA 1513 1186 139 178 ...
其中labels 是由Unicode，x，y，w，h组成的


In [24]:
# 定义一个空的DataFrame，用于存储kaggle submit的格式
kaggle_submit_df = pd.DataFrame(columns=["image_id","labels"])

# 按照Image分组，对每一组进行处理
for image_id, group in merged_test_df.groupby("Image"):
    # 初始化一个空的字符串，用于存储labels
    labels = ""
    # 遍历每一行，获取Unicode，x，y，width，height
    for index, row in group.iterrows():
        unicode = row["Unicode"]
        x = row["X"]
        y = row["Y"]
        width = row["Width"]
        height = row["Height"]
        # 把Unicode，x，y，width，height拼接成一个字符串，并添加到labels中
        labels += "{} {} {} {} {} ".format(unicode,x,y,width,height)
    # 去掉labels最后的空格
    labels = labels.strip()
    # 把image_id和labels添加到kaggle_submit_df中
    kaggle_submit_df = pd.concat([kaggle_submit_df,pd.DataFrame({"image_id":[image_id],"labels":[labels]})],ignore_index=True)


# 打印kaggle_submit_df
print(kaggle_submit_df)

KeyboardInterrupt: 

In [ ]:
# 将合并后的DataFrame保存为一个新的csv文件
kaggle_submit_df.to_csv("ground_truth.csv", index=False)

## 检查那张错误图片修了没有，逆天
修好了！好消息

In [ ]:
# 导入需要的库
import cv2
import numpy as np

# 定义图片和标注文件的路径
img_path = "200004148_00074_2.jpg"
ann_path = "WenTi.txt"

# 读取图片
img = cv2.imread(img_path)

# 读取标注文件，每一行是一个标注信息
with open(ann_path, "r") as f:
    lines = f.readlines()

# 遍历每一行，解析标注信息
for line in lines:
    # 去掉换行符，按逗号分割
    line = line.strip().split(",")
    # 获取Unicode，x，y，width，height
    unicode = line[0]
    x = int(line[2])
    y = int(line[3])
    width = int(line[6])
    height = int(line[7])
    # 计算左上角和右下角的坐标
    xmin = x
    ymin = y
    xmax = x + width
    ymax = y + height
    # 在图片上画出矩形框，颜色为红色，线宽为2
    cv2.rectangle(img, (xmin, ymin), (xmax, ymax), (0, 0, 255), 2)
    # 在图片上显示Unicode，字体为cv2.FONT_HERSHEY_SIMPLEX，大小为0.5，颜色为白色，线宽为1
    cv2.putText(img, unicode, (xmin, ymin - 5), cv2.FONT_HERSHEY_SIMPLEX, 0.5, (255, 255, 255), 1)

# 显示图片
cv2.imshow("Image with annotations", img)
cv2.waitKey(0)
cv2.destroyAllWindows()

In [ ]:
# 导入COCO API
from pycocotools.coco import COCO

# 指定数据集和标注文件的路径
dataDir = "test"
annFile = "{}.json".format(dataDir)

# 创建COCO对象，加载标注文件
coco = COCO(annFile)

# 获取所有的图像id
imgIds = coco.getImgIds()

# 获取所有的类别id和名称
catIds = coco.getCatIds()
catNms = coco.loadCats(catIds)

# 打印一些信息
print("Number of images: {}".format(len(imgIds)))
print("Number of categories: {}".format(len(catIds)))
print("Category names: {}".format([cat["name"] for cat in catNms]))

# 4. 训练集处理
流程和测试集一样，也不需要删除那张照片了，喜大普奔。

### 首先依然是合并所有csv

In [25]:
import pandas as pd
import os

# 定义数据集的根目录
root_dir = r"train_origin"

# 创建一个空的DataFrame来存储合并后的数据
merged_train_df = pd.DataFrame()

# 遍历根目录下的所有文件夹，每个文件夹对应一本kuzushiji书
for folder in os.listdir(root_dir):
    # 获取当前文件夹的路径
    folder_path = os.path.join(root_dir, folder)
    # 获取当前文件夹下的csv文件的路径
    csv_path = os.path.join(folder_path, folder + "_coordinate.csv")
    # 读取csv文件为一个DataFrame
    csv_df = pd.read_csv(csv_path)
    # 删去Block ID和Char ID两列
    csv_df.drop(columns=["Block ID", "Char ID"], inplace=True)
    # 将当前DataFrame追加到合并后的DataFrame中
    merged_train_df = pd.concat([merged_train_df, csv_df], ignore_index=True) # 修改这一行

# 将合并后的DataFrame保存为一个新的csv文件
merged_train_df.to_csv("merged_train_coordinate.csv", index=False)

In [26]:
merged_train_df

Unicode              Image     X     Y  Width  Height
0       U+81EA  100241706_00004_2  1852  1736    104     219
1       U+5E8F  100241706_00004_2  1816  2096    152     296
2       U+82E5  100241706_00004_2  1465   951    172     216
3       U+3044  100241706_00004_2  1495  1218    143      69
4       U+6642  100241706_00004_2  1465  1338    168     179
...        ...                ...   ...   ...    ...     ...
684104  U+306E        umgy012_042   139  1343     63      73
684105  U+5341        umgy012_042   120  1429     89      83
684106  U+4E8C        umgy012_042   119  1514    105      48
684107  U+5927        umgy012_042   145  1574    111      83
684108  U+5C3E        umgy012_042   135  1657    113      92

[684109 rows x 6 columns]

In [27]:
# 提取Image列为一个Series
train_series = merged_train_df["Image"]

# 删除重复的值，并转换为一个列表
train_list = train_series.drop_duplicates().tolist()

# 定义train_images的文件夹路径
train_images_dir = "train"

train_list

['100241706_00004_2',
 '100241706_00005_1',
 '100241706_00005_2',
 '100241706_00006_1',
 '100241706_00007_2',
 '100241706_00008_1',
 '100241706_00008_2',
 '100241706_00009_1',
 '100241706_00009_2',
 '100241706_00010_1',
 '100241706_00010_2',
 '100241706_00011_1',
 '100241706_00011_2',
 '100241706_00012_1',
 '100241706_00012_2',
 '100241706_00013_1',
 '100241706_00013_2',
 '100241706_00014_1',
 '100241706_00014_2',
 '100241706_00015_1',
 '100241706_00015_2',
 '100241706_00016_1',
 '100241706_00016_2',
 '100241706_00017_1',
 '100241706_00017_2',
 '100241706_00018_1',
 '100241706_00018_2',
 '100241706_00019_1',
 '100241706_00019_2',
 '100241706_00020_1',
 '100241706_00020_2',
 '100241706_00021_1',
 '100241706_00021_2',
 '100241706_00022_1',
 '100241706_00022_2',
 '100241706_00023_1',
 '100241706_00023_2',
 '100241706_00024_1',
 '100241706_00024_2',
 '100241706_00025_1',
 '100241706_00025_2',
 '100241706_00026_1',
 '100241706_00026_2',
 '100241706_00027_1',
 '100241706_00027_2',
 '10024170

### 从分列的文件下提取训练集照片

In [28]:
import os
import shutil

source = "train_origin" # 源文件夹路径
dest = "train" # 目标文件夹路径
if not os.path.exists (dest): # 如果目标文件夹不存在，就创建一个
    os.mkdir (dest)

for folder in os.listdir (source): # 遍历源文件夹下的所有子文件夹
    folder_path = os.path.join (source, folder) # 获取子文件夹的完整路径
    if os.path.isdir (folder_path): # 如果是文件夹，就继续遍历
        images_path = os.path.join (folder_path, "images") # 获取images文件夹的完整路径
        for image in os.listdir (images_path): # 遍历images文件夹下的所有图片
            image_path = os.path.join (images_path, image) # 获取图片的完整路径
            shutil.copy (image_path, dest) # 把图片复制到目标文件夹下
            print ("已复制", image_path, "到", dest)


已复制 train_origin\100241706\images\100241706_00004_2.jpg 到 train
已复制 train_origin\100241706\images\100241706_00005_1.jpg 到 train
已复制 train_origin\100241706\images\100241706_00005_2.jpg 到 train
已复制 train_origin\100241706\images\100241706_00006_1.jpg 到 train
已复制 train_origin\100241706\images\100241706_00007_2.jpg 到 train
已复制 train_origin\100241706\images\100241706_00008_1.jpg 到 train
已复制 train_origin\100241706\images\100241706_00008_2.jpg 到 train
已复制 train_origin\100241706\images\100241706_00009_1.jpg 到 train
已复制 train_origin\100241706\images\100241706_00009_2.jpg 到 train
已复制 train_origin\100241706\images\100241706_00010_1.jpg 到 train
已复制 train_origin\100241706\images\100241706_00010_2.jpg 到 train
已复制 train_origin\100241706\images\100241706_00011_1.jpg 到 train
已复制 train_origin\100241706\images\100241706_00011_2.jpg 到 train
已复制 train_origin\100241706\images\100241706_00012_1.jpg 到 train
已复制 train_origin\100241706\images\100241706_00012_2.jpg 到 train
已复制 train_origin\100241706\images\100241

已复制 train_origin\100249376\images\100249376_00005_1.jpg 到 train
已复制 train_origin\100249376\images\100249376_00005_2.jpg 到 train
已复制 train_origin\100249376\images\100249376_00006_1.jpg 到 train
已复制 train_origin\100249376\images\100249376_00006_2.jpg 到 train
已复制 train_origin\100249376\images\100249376_00007_1.jpg 到 train
已复制 train_origin\100249376\images\100249376_00007_2.jpg 到 train
已复制 train_origin\100249376\images\100249376_00008_1.jpg 到 train
已复制 train_origin\100249376\images\100249376_00008_2.jpg 到 train
已复制 train_origin\100249376\images\100249376_00009_1.jpg 到 train
已复制 train_origin\100249376\images\100249376_00009_2.jpg 到 train
已复制 train_origin\100249376\images\100249376_00010_1.jpg 到 train
已复制 train_origin\100249376\images\100249376_00010_2.jpg 到 train
已复制 train_origin\100249376\images\100249376_00011_1.jpg 到 train
已复制 train_origin\100249376\images\100249376_00011_2.jpg 到 train
已复制 train_origin\100249376\images\100249376_00012_1.jpg 到 train
已复制 train_origin\100249376\images\100249

已复制 train_origin\100249416\images\100249416_00030_2.jpg 到 train
已复制 train_origin\100249416\images\100249416_00031_1.jpg 到 train
已复制 train_origin\100249416\images\100249416_00031_2.jpg 到 train
已复制 train_origin\100249416\images\100249416_00032_1.jpg 到 train
已复制 train_origin\100249416\images\100249416_00032_2.jpg 到 train
已复制 train_origin\100249416\images\100249416_00033_1.jpg 到 train
已复制 train_origin\100249416\images\100249416_00033_2.jpg 到 train
已复制 train_origin\100249416\images\100249416_00034_1.jpg 到 train
已复制 train_origin\100249476\images\100249476_00001_1.jpg 到 train
已复制 train_origin\100249476\images\100249476_00002_2.jpg 到 train
已复制 train_origin\100249476\images\100249476_00003_1.jpg 到 train
已复制 train_origin\100249476\images\100249476_00003_2.jpg 到 train
已复制 train_origin\100249476\images\100249476_00004_1.jpg 到 train
已复制 train_origin\100249476\images\100249476_00004_2.jpg 到 train
已复制 train_origin\100249476\images\100249476_00006_1.jpg 到 train
已复制 train_origin\100249476\images\100249

已复制 train_origin\100249537\images\100249537_00063_1.jpg 到 train
已复制 train_origin\100249537\images\100249537_00063_2.jpg 到 train
已复制 train_origin\100249537\images\100249537_00064_1.jpg 到 train
已复制 train_origin\100249537\images\100249537_00064_2.jpg 到 train
已复制 train_origin\100249537\images\100249537_00065_1.jpg 到 train
已复制 train_origin\100249537\images\100249537_00065_2.jpg 到 train
已复制 train_origin\100249537\images\100249537_00066_1.jpg 到 train
已复制 train_origin\100249537\images\100249537_00066_2.jpg 到 train
已复制 train_origin\100249537\images\100249537_00067_1.jpg 到 train
已复制 train_origin\100249537\images\100249537_00067_2.jpg 到 train
已复制 train_origin\100249537\images\100249537_00068_1.jpg 到 train
已复制 train_origin\100249537\images\100249537_00068_2.jpg 到 train
已复制 train_origin\100249537\images\100249537_00069_1.jpg 到 train
已复制 train_origin\100249537\images\100249537_00069_2.jpg 到 train
已复制 train_origin\100249537\images\100249537_00072_2.jpg 到 train
已复制 train_origin\100249537\images\100249

已复制 train_origin\200003076\images\200003076_00074_1.jpg 到 train
已复制 train_origin\200003076\images\200003076_00074_2.jpg 到 train
已复制 train_origin\200003076\images\200003076_00075_1.jpg 到 train
已复制 train_origin\200003076\images\200003076_00075_2.jpg 到 train
已复制 train_origin\200003076\images\200003076_00076_1.jpg 到 train
已复制 train_origin\200003076\images\200003076_00077_1.jpg 到 train
已复制 train_origin\200003076\images\200003076_00077_2.jpg 到 train
已复制 train_origin\200003076\images\200003076_00078_1.jpg 到 train
已复制 train_origin\200003076\images\200003076_00078_2.jpg 到 train
已复制 train_origin\200003076\images\200003076_00079_1.jpg 到 train
已复制 train_origin\200003076\images\200003076_00080_1.jpg 到 train
已复制 train_origin\200003076\images\200003076_00080_2.jpg 到 train
已复制 train_origin\200003076\images\200003076_00081_1.jpg 到 train
已复制 train_origin\200003076\images\200003076_00081_2.jpg 到 train
已复制 train_origin\200003076\images\200003076_00082_1.jpg 到 train
已复制 train_origin\200003076\images\200003

已复制 train_origin\200003076\images\200003076_00169_2.jpg 到 train
已复制 train_origin\200003076\images\200003076_00170_1.jpg 到 train
已复制 train_origin\200003076\images\200003076_00170_2.jpg 到 train
已复制 train_origin\200003076\images\200003076_00171_1.jpg 到 train
已复制 train_origin\200003076\images\200003076_00172_1.jpg 到 train
已复制 train_origin\200003076\images\200003076_00172_2.jpg 到 train
已复制 train_origin\200003076\images\200003076_00173_1.jpg 到 train
已复制 train_origin\200003076\images\200003076_00173_2.jpg 到 train
已复制 train_origin\200003076\images\200003076_00174_1.jpg 到 train
已复制 train_origin\200003076\images\200003076_00175_2.jpg 到 train
已复制 train_origin\200003076\images\200003076_00176_1.jpg 到 train
已复制 train_origin\200003076\images\200003076_00176_2.jpg 到 train
已复制 train_origin\200003076\images\200003076_00177_1.jpg 到 train
已复制 train_origin\200003967\images\200003967_00004_2.jpg 到 train
已复制 train_origin\200003967\images\200003967_00005_1.jpg 到 train
已复制 train_origin\200003967\images\200003

已复制 train_origin\200004148\images\200004148_00054_2.jpg 到 train
已复制 train_origin\200004148\images\200004148_00055_1.jpg 到 train
已复制 train_origin\200004148\images\200004148_00055_2.jpg 到 train
已复制 train_origin\200004148\images\200004148_00056_1.jpg 到 train
已复制 train_origin\200004148\images\200004148_00056_2.jpg 到 train
已复制 train_origin\200004148\images\200004148_00058_1.jpg 到 train
已复制 train_origin\200004148\images\200004148_00058_2.jpg 到 train
已复制 train_origin\200004148\images\200004148_00060_1.jpg 到 train
已复制 train_origin\200004148\images\200004148_00060_2.jpg 到 train
已复制 train_origin\200004148\images\200004148_00061_1.jpg 到 train
已复制 train_origin\200004148\images\200004148_00061_2.jpg 到 train
已复制 train_origin\200004148\images\200004148_00062_1.jpg 到 train
已复制 train_origin\200004148\images\200004148_00062_2.jpg 到 train
已复制 train_origin\200004148\images\200004148_00064_1.jpg 到 train
已复制 train_origin\200004148\images\200004148_00064_2.jpg 到 train
已复制 train_origin\200004148\images\200004

已复制 train_origin\200014685\images\200014685_00006_1.jpg 到 train
已复制 train_origin\200014685\images\200014685_00006_2.jpg 到 train
已复制 train_origin\200014685\images\200014685_00007_1.jpg 到 train
已复制 train_origin\200014685\images\200014685_00007_2.jpg 到 train
已复制 train_origin\200014685\images\200014685_00008_1.jpg 到 train
已复制 train_origin\200014685\images\200014685_00008_2.jpg 到 train
已复制 train_origin\200014685\images\200014685_00009_1.jpg 到 train
已复制 train_origin\200014685\images\200014685_00009_2.jpg 到 train
已复制 train_origin\200014685\images\200014685_00010_1.jpg 到 train
已复制 train_origin\200014685\images\200014685_00010_2.jpg 到 train
已复制 train_origin\200014685\images\200014685_00011_1.jpg 到 train
已复制 train_origin\200014685\images\200014685_00011_2.jpg 到 train
已复制 train_origin\200014685\images\200014685_00012_1.jpg 到 train
已复制 train_origin\200014685\images\200014685_00012_2.jpg 到 train
已复制 train_origin\200014685\images\200014685_00013_1.jpg 到 train
已复制 train_origin\200014685\images\200014

已复制 train_origin\200014740\images\200014740_00068_1.jpg 到 train
已复制 train_origin\200014740\images\200014740_00068_2.jpg 到 train
已复制 train_origin\200014740\images\200014740_00069_1.jpg 到 train
已复制 train_origin\200014740\images\200014740_00069_2.jpg 到 train
已复制 train_origin\200014740\images\200014740_00070_1.jpg 到 train
已复制 train_origin\200014740\images\200014740_00070_2.jpg 到 train
已复制 train_origin\200014740\images\200014740_00071_1.jpg 到 train
已复制 train_origin\200014740\images\200014740_00071_2.jpg 到 train
已复制 train_origin\200014740\images\200014740_00073_1.jpg 到 train
已复制 train_origin\200014740\images\200014740_00073_2.jpg 到 train
已复制 train_origin\200014740\images\200014740_00074_1.jpg 到 train
已复制 train_origin\200014740\images\200014740_00074_2.jpg 到 train
已复制 train_origin\200014740\images\200014740_00075_1.jpg 到 train
已复制 train_origin\200014740\images\200014740_00075_2.jpg 到 train
已复制 train_origin\200014740\images\200014740_00076_2.jpg 到 train
已复制 train_origin\200014740\images\200014

已复制 train_origin\200015779\images\200015779_00084_1.jpg 到 train
已复制 train_origin\200015779\images\200015779_00084_2.jpg 到 train
已复制 train_origin\200015779\images\200015779_00085_1.jpg 到 train
已复制 train_origin\200015779\images\200015779_00085_2.jpg 到 train
已复制 train_origin\200015779\images\200015779_00088_1.jpg 到 train
已复制 train_origin\200015779\images\200015779_00088_2.jpg 到 train
已复制 train_origin\200015779\images\200015779_00089_2.jpg 到 train
已复制 train_origin\200015779\images\200015779_00090_1.jpg 到 train
已复制 train_origin\200015779\images\200015779_00090_2.jpg 到 train
已复制 train_origin\200015779\images\200015779_00091_1.jpg 到 train
已复制 train_origin\200015779\images\200015779_00091_2.jpg 到 train
已复制 train_origin\200015779\images\200015779_00092_1.jpg 到 train
已复制 train_origin\200015779\images\200015779_00092_2.jpg 到 train
已复制 train_origin\200015779\images\200015779_00093_1.jpg 到 train
已复制 train_origin\200015779\images\200015779_00093_2.jpg 到 train
已复制 train_origin\200015779\images\200015

已复制 train_origin\200021637\images\200021637_00008_1.jpg 到 train
已复制 train_origin\200021637\images\200021637_00008_2.jpg 到 train
已复制 train_origin\200021637\images\200021637_00009_1.jpg 到 train
已复制 train_origin\200021637\images\200021637_00009_2.jpg 到 train
已复制 train_origin\200021637\images\200021637_00010_1.jpg 到 train
已复制 train_origin\200021637\images\200021637_00010_2.jpg 到 train
已复制 train_origin\200021637\images\200021637_00011_1.jpg 到 train
已复制 train_origin\200021637\images\200021637_00011_2.jpg 到 train
已复制 train_origin\200021637\images\200021637_00012_1.jpg 到 train
已复制 train_origin\200021637\images\200021637_00012_2.jpg 到 train
已复制 train_origin\200021637\images\200021637_00013_1.jpg 到 train
已复制 train_origin\200021637\images\200021637_00013_2.jpg 到 train
已复制 train_origin\200021637\images\200021637_00014_1.jpg 到 train
已复制 train_origin\200021637\images\200021637_00014_2.jpg 到 train
已复制 train_origin\200021637\images\200021637_00015_1.jpg 到 train
已复制 train_origin\200021637\images\200021

已复制 train_origin\200021660\images\200021660_00031_1.jpg 到 train
已复制 train_origin\200021660\images\200021660_00031_2.jpg 到 train
已复制 train_origin\200021660\images\200021660_00032_1.jpg 到 train
已复制 train_origin\200021660\images\200021660_00032_2.jpg 到 train
已复制 train_origin\200021660\images\200021660_00033_1.jpg 到 train
已复制 train_origin\200021660\images\200021660_00036_2.jpg 到 train
已复制 train_origin\200021660\images\200021660_00037_1.jpg 到 train
已复制 train_origin\200021660\images\200021660_00037_2.jpg 到 train
已复制 train_origin\200021660\images\200021660_00038_1.jpg 到 train
已复制 train_origin\200021660\images\200021660_00038_2.jpg 到 train
已复制 train_origin\200021660\images\200021660_00039_1.jpg 到 train
已复制 train_origin\200021660\images\200021660_00039_2.jpg 到 train
已复制 train_origin\200021660\images\200021660_00040_1.jpg 到 train
已复制 train_origin\200021660\images\200021660_00040_2.jpg 到 train
已复制 train_origin\200021660\images\200021660_00041_1.jpg 到 train
已复制 train_origin\200021660\images\200021

已复制 train_origin\200021712\images\200021712_00016_1.jpg 到 train
已复制 train_origin\200021712\images\200021712_00016_2.jpg 到 train
已复制 train_origin\200021712\images\200021712_00017_1.jpg 到 train
已复制 train_origin\200021712\images\200021712_00017_2.jpg 到 train
已复制 train_origin\200021712\images\200021712_00018_1.jpg 到 train
已复制 train_origin\200021712\images\200021712_00018_2.jpg 到 train
已复制 train_origin\200021712\images\200021712_00019_1.jpg 到 train
已复制 train_origin\200021712\images\200021712_00019_2.jpg 到 train
已复制 train_origin\200021712\images\200021712_00020_1.jpg 到 train
已复制 train_origin\200021712\images\200021712_00020_2.jpg 到 train
已复制 train_origin\200021712\images\200021712_00021_1.jpg 到 train
已复制 train_origin\200021712\images\200021712_00021_2.jpg 到 train
已复制 train_origin\200021712\images\200021712_00022_1.jpg 到 train
已复制 train_origin\200021712\images\200021712_00022_2.jpg 到 train
已复制 train_origin\200021712\images\200021712_00023_1.jpg 到 train
已复制 train_origin\200021712\images\200021

已复制 train_origin\200021763\images\200021763_00007_1.jpg 到 train
已复制 train_origin\200021763\images\200021763_00007_2.jpg 到 train
已复制 train_origin\200021763\images\200021763_00008_1.jpg 到 train
已复制 train_origin\200021763\images\200021763_00008_2.jpg 到 train
已复制 train_origin\200021763\images\200021763_00009_1.jpg 到 train
已复制 train_origin\200021763\images\200021763_00009_2.jpg 到 train
已复制 train_origin\200021763\images\200021763_00010_1.jpg 到 train
已复制 train_origin\200021763\images\200021763_00010_2.jpg 到 train
已复制 train_origin\200021763\images\200021763_00011_1.jpg 到 train
已复制 train_origin\200021763\images\200021763_00011_2.jpg 到 train
已复制 train_origin\200021763\images\200021763_00012_1.jpg 到 train
已复制 train_origin\200021763\images\200021763_00012_2.jpg 到 train
已复制 train_origin\200021763\images\200021763_00013_1.jpg 到 train
已复制 train_origin\200021763\images\200021763_00013_2.jpg 到 train
已复制 train_origin\200021763\images\200021763_00014_1.jpg 到 train
已复制 train_origin\200021763\images\200021

已复制 train_origin\200021802\images\200021802_00029_1.jpg 到 train
已复制 train_origin\200021802\images\200021802_00029_2.jpg 到 train
已复制 train_origin\200021802\images\200021802_00030_1.jpg 到 train
已复制 train_origin\200021802\images\200021802_00030_2.jpg 到 train
已复制 train_origin\200021802\images\200021802_00031_1.jpg 到 train
已复制 train_origin\200021802\images\200021802_00031_2.jpg 到 train
已复制 train_origin\200021802\images\200021802_00032_1.jpg 到 train
已复制 train_origin\200021802\images\200021802_00032_2.jpg 到 train
已复制 train_origin\200021802\images\200021802_00033_1.jpg 到 train
已复制 train_origin\200021802\images\200021802_00033_2.jpg 到 train
已复制 train_origin\200021802\images\200021802_00034_1.jpg 到 train
已复制 train_origin\200021802\images\200021802_00034_2.jpg 到 train
已复制 train_origin\200021802\images\200021802_00035_1.jpg 到 train
已复制 train_origin\200021802\images\200021802_00035_2.jpg 到 train
已复制 train_origin\200021802\images\200021802_00036_1.jpg 到 train
已复制 train_origin\200021802\images\200021

已复制 train_origin\200021853\images\200021853_00021_2.jpg 到 train
已复制 train_origin\200021853\images\200021853_00022_1.jpg 到 train
已复制 train_origin\200021853\images\200021853_00022_2.jpg 到 train
已复制 train_origin\200021853\images\200021853_00023_1.jpg 到 train
已复制 train_origin\200021853\images\200021853_00023_2.jpg 到 train
已复制 train_origin\200021853\images\200021853_00024_1.jpg 到 train
已复制 train_origin\200021853\images\200021853_00024_2.jpg 到 train
已复制 train_origin\200021853\images\200021853_00025_1.jpg 到 train
已复制 train_origin\200021853\images\200021853_00025_2.jpg 到 train
已复制 train_origin\200021853\images\200021853_00026_1.jpg 到 train
已复制 train_origin\200021853\images\200021853_00026_2.jpg 到 train
已复制 train_origin\200021853\images\200021853_00027_1.jpg 到 train
已复制 train_origin\200021853\images\200021853_00027_2.jpg 到 train
已复制 train_origin\200021853\images\200021853_00028_1.jpg 到 train
已复制 train_origin\200021853\images\200021853_00028_2.jpg 到 train
已复制 train_origin\200021853\images\200021

已复制 train_origin\brsk00000\images\brsk001_033.jpg 到 train
已复制 train_origin\brsk00000\images\brsk001_034.jpg 到 train
已复制 train_origin\brsk00000\images\brsk001_035.jpg 到 train
已复制 train_origin\brsk00000\images\brsk001_036.jpg 到 train
已复制 train_origin\brsk00000\images\brsk002_003.jpg 到 train
已复制 train_origin\brsk00000\images\brsk002_004.jpg 到 train
已复制 train_origin\brsk00000\images\brsk002_005.jpg 到 train
已复制 train_origin\brsk00000\images\brsk002_006.jpg 到 train
已复制 train_origin\brsk00000\images\brsk002_007.jpg 到 train
已复制 train_origin\brsk00000\images\brsk002_008.jpg 到 train
已复制 train_origin\brsk00000\images\brsk002_009.jpg 到 train
已复制 train_origin\brsk00000\images\brsk002_010.jpg 到 train
已复制 train_origin\brsk00000\images\brsk002_011.jpg 到 train
已复制 train_origin\brsk00000\images\brsk002_012.jpg 到 train
已复制 train_origin\brsk00000\images\brsk002_013.jpg 到 train
已复制 train_origin\brsk00000\images\brsk002_014.jpg 到 train
已复制 train_origin\brsk00000\images\brsk002_015.jpg 到 train
已复制 train_orig

已复制 train_origin\brsk00000\images\brsk005_028.jpg 到 train
已复制 train_origin\brsk00000\images\brsk005_029.jpg 到 train
已复制 train_origin\brsk00000\images\brsk005_030.jpg 到 train
已复制 train_origin\brsk00000\images\brsk005_031.jpg 到 train
已复制 train_origin\brsk00000\images\brsk005_032.jpg 到 train
已复制 train_origin\brsk00000\images\brsk005_033.jpg 到 train
已复制 train_origin\brsk00000\images\brsk005_034.jpg 到 train
已复制 train_origin\brsk00000\images\brsk005_035.jpg 到 train
已复制 train_origin\brsk00000\images\brsk005_036.jpg 到 train
已复制 train_origin\brsk00000\images\brsk005_037.jpg 到 train
已复制 train_origin\brsk00000\images\brsk005_038.jpg 到 train
已复制 train_origin\brsk00000\images\brsk005_039.jpg 到 train
已复制 train_origin\brsk00000\images\brsk005_040.jpg 到 train
已复制 train_origin\brsk00000\images\brsk005_041.jpg 到 train
已复制 train_origin\brsk00000\images\brsk005_042.jpg 到 train
已复制 train_origin\hnsd00000\images\hnsd001_003.jpg 到 train
已复制 train_origin\hnsd00000\images\hnsd001_004.jpg 到 train
已复制 train_orig

已复制 train_origin\hnsd00000\images\hnsd004_024.jpg 到 train
已复制 train_origin\hnsd00000\images\hnsd004_025.jpg 到 train
已复制 train_origin\hnsd00000\images\hnsd004_026.jpg 到 train
已复制 train_origin\hnsd00000\images\hnsd004_027.jpg 到 train
已复制 train_origin\hnsd00000\images\hnsd004_028.jpg 到 train
已复制 train_origin\hnsd00000\images\hnsd004_029.jpg 到 train
已复制 train_origin\hnsd00000\images\hnsd004_030.jpg 到 train
已复制 train_origin\hnsd00000\images\hnsd004_031.jpg 到 train
已复制 train_origin\hnsd00000\images\hnsd004_032.jpg 到 train
已复制 train_origin\hnsd00000\images\hnsd004_033.jpg 到 train
已复制 train_origin\hnsd00000\images\hnsd004_034.jpg 到 train
已复制 train_origin\hnsd00000\images\hnsd004_035.jpg 到 train
已复制 train_origin\hnsd00000\images\hnsd004_036.jpg 到 train
已复制 train_origin\hnsd00000\images\hnsd004_037.jpg 到 train
已复制 train_origin\hnsd00000\images\hnsd004_038.jpg 到 train
已复制 train_origin\hnsd00000\images\hnsd004_039.jpg 到 train
已复制 train_origin\hnsd00000\images\hnsd004_040.jpg 到 train
已复制 train_orig

已复制 train_origin\hnsd00000\images\hnsd008_017.jpg 到 train
已复制 train_origin\hnsd00000\images\hnsd008_018.jpg 到 train
已复制 train_origin\hnsd00000\images\hnsd008_019.jpg 到 train
已复制 train_origin\hnsd00000\images\hnsd008_020.jpg 到 train
已复制 train_origin\hnsd00000\images\hnsd008_021.jpg 到 train
已复制 train_origin\hnsd00000\images\hnsd008_022.jpg 到 train
已复制 train_origin\hnsd00000\images\hnsd008_023.jpg 到 train
已复制 train_origin\hnsd00000\images\hnsd008_024.jpg 到 train
已复制 train_origin\hnsd00000\images\hnsd008_025.jpg 到 train
已复制 train_origin\hnsd00000\images\hnsd008_026.jpg 到 train
已复制 train_origin\hnsd00000\images\hnsd008_027.jpg 到 train
已复制 train_origin\hnsd00000\images\hnsd008_028.jpg 到 train
已复制 train_origin\hnsd00000\images\hnsd008_029.jpg 到 train
已复制 train_origin\hnsd00000\images\hnsd008_030.jpg 到 train
已复制 train_origin\hnsd00000\images\hnsd008_031.jpg 到 train
已复制 train_origin\hnsd00000\images\hnsd008_032.jpg 到 train
已复制 train_origin\hnsd00000\images\hnsd008_033.jpg 到 train
已复制 train_orig

已复制 train_origin\hnsd00000\images\hnsd012_026.jpg 到 train
已复制 train_origin\hnsd00000\images\hnsd012_027.jpg 到 train
已复制 train_origin\hnsd00000\images\hnsd012_028.jpg 到 train
已复制 train_origin\hnsd00000\images\hnsd012_029.jpg 到 train
已复制 train_origin\hnsd00000\images\hnsd012_030.jpg 到 train
已复制 train_origin\hnsd00000\images\hnsd012_031.jpg 到 train
已复制 train_origin\hnsd00000\images\hnsd012_032.jpg 到 train
已复制 train_origin\hnsd00000\images\hnsd012_033.jpg 到 train
已复制 train_origin\hnsd00000\images\hnsd012_034.jpg 到 train
已复制 train_origin\hnsd00000\images\hnsd012_035.jpg 到 train
已复制 train_origin\hnsd00000\images\hnsd012_036.jpg 到 train
已复制 train_origin\hnsd00000\images\hnsd012_037.jpg 到 train
已复制 train_origin\hnsd00000\images\hnsd012_038.jpg 到 train
已复制 train_origin\umgy00000\images\umgy001_003.jpg 到 train
已复制 train_origin\umgy00000\images\umgy001_004.jpg 到 train
已复制 train_origin\umgy00000\images\umgy001_005.jpg 到 train
已复制 train_origin\umgy00000\images\umgy001_006.jpg 到 train
已复制 train_orig

已复制 train_origin\umgy00000\images\umgy004_034.jpg 到 train
已复制 train_origin\umgy00000\images\umgy004_035.jpg 到 train
已复制 train_origin\umgy00000\images\umgy004_036.jpg 到 train
已复制 train_origin\umgy00000\images\umgy004_037.jpg 到 train
已复制 train_origin\umgy00000\images\umgy004_038.jpg 到 train
已复制 train_origin\umgy00000\images\umgy004_039.jpg 到 train
已复制 train_origin\umgy00000\images\umgy004_040.jpg 到 train
已复制 train_origin\umgy00000\images\umgy004_041.jpg 到 train
已复制 train_origin\umgy00000\images\umgy004_042.jpg 到 train
已复制 train_origin\umgy00000\images\umgy004_043.jpg 到 train
已复制 train_origin\umgy00000\images\umgy004_044.jpg 到 train
已复制 train_origin\umgy00000\images\umgy004_045.jpg 到 train
已复制 train_origin\umgy00000\images\umgy004_046.jpg 到 train
已复制 train_origin\umgy00000\images\umgy005_001.jpg 到 train
已复制 train_origin\umgy00000\images\umgy005_002.jpg 到 train
已复制 train_origin\umgy00000\images\umgy005_003.jpg 到 train
已复制 train_origin\umgy00000\images\umgy005_004.jpg 到 train
已复制 train_orig

已复制 train_origin\umgy00000\images\umgy009_023.jpg 到 train
已复制 train_origin\umgy00000\images\umgy009_024.jpg 到 train
已复制 train_origin\umgy00000\images\umgy009_025.jpg 到 train
已复制 train_origin\umgy00000\images\umgy009_026.jpg 到 train
已复制 train_origin\umgy00000\images\umgy009_027.jpg 到 train
已复制 train_origin\umgy00000\images\umgy009_028.jpg 到 train
已复制 train_origin\umgy00000\images\umgy009_029.jpg 到 train
已复制 train_origin\umgy00000\images\umgy009_030.jpg 到 train
已复制 train_origin\umgy00000\images\umgy009_031.jpg 到 train
已复制 train_origin\umgy00000\images\umgy009_032.jpg 到 train
已复制 train_origin\umgy00000\images\umgy009_033.jpg 到 train
已复制 train_origin\umgy00000\images\umgy009_034.jpg 到 train
已复制 train_origin\umgy00000\images\umgy009_035.jpg 到 train
已复制 train_origin\umgy00000\images\umgy009_036.jpg 到 train
已复制 train_origin\umgy00000\images\umgy009_037.jpg 到 train
已复制 train_origin\umgy00000\images\umgy009_038.jpg 到 train
已复制 train_origin\umgy00000\images\umgy009_039.jpg 到 train
已复制 train_orig

一共3,629张照片

### 生成images字段

In [29]:
# 创建一个空的列表来存储coco数据集格式的images字段
coco_train = []

# 遍历image_list中的每个照片名
for image in train_list:
    # 获取当前照片文件的路径
    image_path = os.path.join(train_images_dir, image+'.jpg')
    # 打开当前照片文件，获取宽度和高度
    img = Image.open(image_path)
    width, height = img.size
    # 关闭当前照片文件
    img.close()
    # 创建一个字典，存储当前照片的信息，包括id，file_name，width和height
    image_info = {
        "file_name": image+'.jpg',
        "id": train_list.index(image), # id从0开始递增
        "height": height,
        "width": width
    }
    # 将字典追加到coco_images列表中
    coco_train.append(image_info)

# 打印结果
print("coco数据集格式的images字段为：", coco_train)

coco数据集格式的images字段为： [{'file_name': '100241706_00004_2.jpg', 'id': 0, 'height': 3874, 'width': 2404}, {'file_name': '100241706_00005_1.jpg', 'id': 1, 'height': 3874, 'width': 2392}, {'file_name': '100241706_00005_2.jpg', 'id': 2, 'height': 3874, 'width': 2416}, {'file_name': '100241706_00006_1.jpg', 'id': 3, 'height': 3886, 'width': 2362}, {'file_name': '100241706_00007_2.jpg', 'id': 4, 'height': 3868, 'width': 2392}, {'file_name': '100241706_00008_1.jpg', 'id': 5, 'height': 3874, 'width': 2350}, {'file_name': '100241706_00008_2.jpg', 'id': 6, 'height': 3868, 'width': 2410}, {'file_name': '100241706_00009_1.jpg', 'id': 7, 'height': 3880, 'width': 2362}, {'file_name': '100241706_00009_2.jpg', 'id': 8, 'height': 3880, 'width': 2374}, {'file_name': '100241706_00010_1.jpg', 'id': 9, 'height': 3872, 'width': 2369}, {'file_name': '100241706_00010_2.jpg', 'id': 10, 'height': 3868, 'width': 2380}, {'file_name': '100241706_00011_1.jpg', 'id': 11, 'height': 3867, 'width': 2350}, {'file_name': '1

In [30]:
len(coco_train)

3629

In [31]:
# 创建一个空的字典来存储images字段中的照片名和id的对应关系
train_image_dict = {}

# 遍历images字段中的每个字典
for image_info in coco_train:
    # 获取当前字典中的照片名和id
    image_name = image_info["file_name"]
    image_id = image_info["id"]
    # 将照片名和id作为键值对存入image_dict中
    train_image_dict[image_name] = image_id

# 创建一个空的字典来存储categories字段中的类别名和id的对应关系
category_dict = {}

# 遍历categories字段中的每个字典
for category_info in categories:
    # 获取当前字典中的类别名和id
    category_name = category_info["name"]
    category_id = category_info["id"]
    # 将类别名和id作为键值对存入category_dict中
    category_dict[category_name] = category_id
    
# 创建一个新的空字典
new_train_image_dict = {}

# 遍历原字典的键值对
for key, value in train_image_dict.items():
    # 去掉键中的.jpg
    new_key = key.replace(".jpg", "")
    # 把新的键值对添加到新字典中
    new_train_image_dict[new_key] = value
new_train_image_dict

{'100241706_00004_2': 0,
 '100241706_00005_1': 1,
 '100241706_00005_2': 2,
 '100241706_00006_1': 3,
 '100241706_00007_2': 4,
 '100241706_00008_1': 5,
 '100241706_00008_2': 6,
 '100241706_00009_1': 7,
 '100241706_00009_2': 8,
 '100241706_00010_1': 9,
 '100241706_00010_2': 10,
 '100241706_00011_1': 11,
 '100241706_00011_2': 12,
 '100241706_00012_1': 13,
 '100241706_00012_2': 14,
 '100241706_00013_1': 15,
 '100241706_00013_2': 16,
 '100241706_00014_1': 17,
 '100241706_00014_2': 18,
 '100241706_00015_1': 19,
 '100241706_00015_2': 20,
 '100241706_00016_1': 21,
 '100241706_00016_2': 22,
 '100241706_00017_1': 23,
 '100241706_00017_2': 24,
 '100241706_00018_1': 25,
 '100241706_00018_2': 26,
 '100241706_00019_1': 27,
 '100241706_00019_2': 28,
 '100241706_00020_1': 29,
 '100241706_00020_2': 30,
 '100241706_00021_1': 31,
 '100241706_00021_2': 32,
 '100241706_00022_1': 33,
 '100241706_00022_2': 34,
 '100241706_00023_1': 35,
 '100241706_00023_2': 36,
 '100241706_00024_1': 37,
 '100241706_00024_2': 

In [32]:
len(new_train_image_dict)

3629

In [34]:
# 创建一个空的列表来存储coco数据集格式的annotations字段
coco_annotations = []

# 遍历merged_df中的每一行
for index, row in merged_train_df.iterrows():
    # 获取当前行的照片名，类别名，坐标和面积
    image_name = row["Image"]
    category_name = row["Unicode"]
    bbox = [row["X"], row["Y"], row["Width"], row["Height"]]
    area = row["Width"] * row["Height"]
    # 创建一个字典，存储当前行的信息，包括id，image_id，category_id，bbox和area
    
    annotation_info = {
        "id": index, # id从0开始递增
        "image_id": new_train_image_dict[image_name], # 根据照片名在new_image_dict中查找对应的id
        "category_id": category_dict[category_name], # 根据类别名在category_dict中查找对应的id
        "bbox": bbox,
        "area": area,
        "iscrowd": 0
    }
    # 将字典追加到coco_annotations列表中
    coco_annotations.append(annotation_info)

In [35]:
coco_annotations

[{'id': 0,
  'image_id': 0,
  'category_id': 2989,
  'bbox': [1852, 1736, 104, 219],
  'area': 22776,
  'iscrowd': 0},
 {'id': 1,
  'image_id': 0,
  'category_id': 1203,
  'bbox': [1816, 2096, 152, 296],
  'area': 44992,
  'iscrowd': 0},
 {'id': 2,
  'image_id': 0,
  'category_id': 3050,
  'bbox': [1465, 951, 172, 216],
  'area': 37152,
  'iscrowd': 0},
 {'id': 3,
  'image_id': 0,
  'category_id': 52,
  'bbox': [1495, 1218, 143, 69],
  'area': 9867,
  'iscrowd': 0},
 {'id': 4,
  'image_id': 0,
  'category_id': 1679,
  'bbox': [1465, 1338, 168, 179],
  'area': 30072,
  'iscrowd': 0},
 {'id': 5,
  'image_id': 0,
  'category_id': 91,
  'bbox': [1497, 1567, 123, 152],
  'area': 18696,
  'iscrowd': 0},
 {'id': 6,
  'image_id': 0,
  'category_id': 2002,
  'bbox': [1504, 1754, 145, 200],
  'area': 29000,
  'iscrowd': 0},
 {'id': 7,
  'image_id': 0,
  'category_id': 1253,
  'bbox': [1479, 2034, 163, 145],
  'area': 23635,
  'iscrowd': 0},
 {'id': 8,
  'image_id': 0,
  'category_id': 88,
  'bbo

In [36]:
train_dict_to_save = {'annotations':coco_annotations, 'images':coco_train, 'categories':categories}

In [37]:
import json
# 日语字符以utf-8编码了，可以正常读取！
with open("train.json", 'w',encoding='utf-8') as f:
    json.dump(train_dict_to_save, f)

### 验证

In [38]:
# 导入COCO API
from pycocotools.coco import COCO

# 指定数据集和标注文件的路径
dataDir = "train"
annFile = "{}.json".format(dataDir)

# 创建COCO对象，加载标注文件
coco = COCO(annFile)

# 获取所有的图像id
imgIds = coco.getImgIds()

# 获取所有的类别id和名称
catIds = coco.getCatIds()
catNms = coco.loadCats(catIds)

# 打印一些信息
print("Number of images: {}".format(len(imgIds)))
print("Number of categories: {}".format(len(catIds)))
print("Category names: {}".format([cat["name"] for cat in catNms]))

loading annotations into memory...
Done (t=2.17s)
creating index...
index created!
Number of images: 3629
Number of categories: 4328
Category names: ['U+0031', 'U+0032', 'U+0034', 'U+0036', 'U+0039', 'U+003F', 'U+2000B', 'U+20D45', 'U+2123D', 'U+22999', 'U+22CE3', 'U+231C3', 'U+2324A', 'U+233D0', 'U+23D0E', 'U+241C6', 'U+242CE', 'U+2437D', 'U+24E30', 'U+24E60', 'U+24FA3', 'U+2564A', 'U+25750', 'U+25877', 'U+25A0', 'U+25A1', 'U+25B2', 'U+25B3', 'U+25CB', 'U+25CF', 'U+25DA1', 'U+25E2E', 'U+25EF', 'U+25FFB', 'U+264B3', 'U+266B0', 'U+27752', 'U+28263', 'U+29780', 'U+29DDA', 'U+29E75', 'U+3002', 'U+3005', 'U+3006', 'U+3007', 'U+300C', 'U+300D', 'U+3031', 'U+3032', 'U+303B', 'U+303D', 'U+3042', 'U+3044', 'U+3046', 'U+3048', 'U+304A', 'U+304B', 'U+304C', 'U+304D', 'U+304E', 'U+304F', 'U+3050', 'U+3051', 'U+3052', 'U+3053', 'U+3054', 'U+3055', 'U+3056', 'U+3057', 'U+3058', 'U+3059', 'U+305A', 'U+305B', 'U+305C', 'U+305D', 'U+305E', 'U+305F', 'U+3060', 'U+3061', 'U+3062', 'U+3063', 'U+3064', 'U

### 生成训练集kaggle格式Ground Truth

In [ ]:
# 定义一个空的DataFrame，用于存储kaggle submit的格式
kaggle_train_df = pd.DataFrame(columns=["image_id","labels"])

# 按照Image分组，对每一组进行处理
for image_id, group in merged_train_df.groupby("Image"):
    # 初始化一个空的字符串，用于存储labels
    labels = ""
    # 遍历每一行，获取Unicode，x，y，width，height
    for index, row in group.iterrows():
        unicode = row["Unicode"]
        x = row["X"]
        y = row["Y"]
        width = row["Width"]
        height = row["Height"]
        # 把Unicode，x，y，width，height拼接成一个字符串，并添加到labels中
        labels += "{} {} {} {} {} ".format(unicode,x,y,width,height)
    # 去掉labels最后的空格
    labels = labels.strip()
    # 把image_id和labels添加到kaggle_train_df中
    kaggle_train_df = pd.concat([kaggle_train_df,pd.DataFrame({"image_id":[image_id],"labels":[labels]})],ignore_index=True)


# 打印kaggle_train_df
print(kaggle_train_df)

In [ ]:
# 将合并后的DataFrame保存为一个新的csv文件
kaggle_train_df.to_csv("train.csv", index=False)

### 验证集

#### 生成kaggle格式

In [39]:
import pandas as pd
val_df = pd.read_csv("200003803_coordinate.csv")
val_df

Unicode              Image     X     Y Block ID Char ID  Width  Height
0      U+3044  200003803_00002_2  1827   554    B0001   C0001    107      77
1      U+3065  200003803_00002_2  1837   642    B0001   C0002    107      88
2      U+308C  200003803_00002_2  1844   745    B0001   C0003     99     106
3      U+306E  200003803_00002_2  1839   867    B0001   C0004     89      83
4      U+5FA1  200003803_00002_2  1841   964    B0001   C0005     91      91
...       ...                ...   ...   ...      ...     ...    ...     ...
11127  U+305F  200003803_00029_2   262  2384    B0001   C0197     49     112
11128  U+308B  200003803_00029_2   274  2496    B0001   C0198     37      91
11129  U+3068  200003803_00029_2   277  2588    B0001   C0199     36     108
11130  U+306A  200003803_00029_2   171  2713    B0001   C0200     72     102
11131  U+3093  200003803_00029_2   176  2821    B0001   C0201     83      97

[11132 rows x 8 columns]

In [ ]:
# 定义一个空的DataFrame，用于存储kaggle submit的格式
kaggle_val_df = pd.DataFrame(columns=["image_id","labels"])

# 按照Image分组，对每一组进行处理
for image_id, group in val_df.groupby("Image"):
    # 初始化一个空的字符串，用于存储labels
    labels = ""
    # 遍历每一行，获取Unicode，x，y，width，height
    for index, row in group.iterrows():
        unicode = row["Unicode"]
        x = row["X"]
        y = row["Y"]
        width = row["Width"]
        height = row["Height"]
        # 把Unicode，x，y，width，height拼接成一个字符串，并添加到labels中
        labels += "{} {} {} {} {} ".format(unicode,x,y,width,height)
    # 去掉labels最后的空格
    labels = labels.strip()
    # 把image_id和labels添加到kaggle_train_df中
    kaggle_val_df = pd.concat([kaggle_val_df,pd.DataFrame({"image_id":[image_id],"labels":[labels]})],ignore_index=True)


# 将合并后的DataFrame保存为一个新的csv文件
kaggle_val_df.to_csv("val.csv", index=False)

In [ ]:
kaggle_val_df.head()

### 生成images字段

In [40]:
# 提取Image列为一个Series
val_series = val_df["Image"]

# 删除重复的值，并转换为一个列表
val_list = val_series.drop_duplicates().tolist()

# 定义val_images的文件夹路径
val_images_dir = "val"

val_list

['200003803_00002_2',
 '200003803_00003_1',
 '200003803_00003_2',
 '200003803_00004_1',
 '200003803_00004_2',
 '200003803_00005_1',
 '200003803_00006_1',
 '200003803_00006_2',
 '200003803_00007_1',
 '200003803_00007_2',
 '200003803_00008_1',
 '200003803_00008_2',
 '200003803_00009_1',
 '200003803_00009_2',
 '200003803_00010_1',
 '200003803_00010_2',
 '200003803_00011_2',
 '200003803_00012_1',
 '200003803_00012_2',
 '200003803_00013_1',
 '200003803_00013_2',
 '200003803_00014_1',
 '200003803_00014_2',
 '200003803_00015_1',
 '200003803_00015_2',
 '200003803_00016_2',
 '200003803_00017_1',
 '200003803_00017_2',
 '200003803_00018_1',
 '200003803_00018_2',
 '200003803_00019_1',
 '200003803_00019_2',
 '200003803_00020_1',
 '200003803_00021_1',
 '200003803_00021_2',
 '200003803_00022_1',
 '200003803_00022_2',
 '200003803_00023_1',
 '200003803_00023_2',
 '200003803_00024_1',
 '200003803_00024_2',
 '200003803_00025_1',
 '200003803_00025_2',
 '200003803_00026_2',
 '200003803_00027_1',
 '20000380

In [41]:
# 创建一个空的列表来存储coco数据集格式的images字段
coco_val = []

# 遍历image_list中的每个照片名
for image in val_list:
    # 获取当前照片文件的路径
    image_path = os.path.join(val_images_dir, image+'.jpg')
    # 打开当前照片文件，获取宽度和高度
    img = Image.open(image_path)
    width, height = img.size
    # 关闭当前照片文件
    img.close()
    # 创建一个字典，存储当前照片的信息，包括id，file_name，width和height
    image_info = {
        "file_name": image+'.jpg',
        "id": val_list.index(image), # id从0开始递增
        "height": height,
        "width": width
    }
    # 将字典追加到coco_images列表中
    coco_val.append(image_info)

# 打印结果
print("coco数据集格式的images字段为：", coco_val)

coco数据集格式的images字段为： [{'file_name': '200003803_00002_2.jpg', 'id': 0, 'height': 3181, 'width': 2159}, {'file_name': '200003803_00003_1.jpg', 'id': 1, 'height': 3187, 'width': 2174}, {'file_name': '200003803_00003_2.jpg', 'id': 2, 'height': 3206, 'width': 2166}, {'file_name': '200003803_00004_1.jpg', 'id': 3, 'height': 3194, 'width': 2154}, {'file_name': '200003803_00004_2.jpg', 'id': 4, 'height': 3202, 'width': 2158}, {'file_name': '200003803_00005_1.jpg', 'id': 5, 'height': 3177, 'width': 2170}, {'file_name': '200003803_00006_1.jpg', 'id': 6, 'height': 3165, 'width': 2138}, {'file_name': '200003803_00006_2.jpg', 'id': 7, 'height': 3191, 'width': 2143}, {'file_name': '200003803_00007_1.jpg', 'id': 8, 'height': 3183, 'width': 2134}, {'file_name': '200003803_00007_2.jpg', 'id': 9, 'height': 3222, 'width': 2139}, {'file_name': '200003803_00008_1.jpg', 'id': 10, 'height': 3211, 'width': 2190}, {'file_name': '200003803_00008_2.jpg', 'id': 11, 'height': 3198, 'width': 2134}, {'file_name': '2

In [42]:
# 创建一个空的字典来存储images字段中的照片名和id的对应关系
val_image_dict = {}

# 遍历images字段中的每个字典
for image_info in coco_val:
    # 获取当前字典中的照片名和id
    image_name = image_info["file_name"]
    image_id = image_info["id"]
    # 将照片名和id作为键值对存入image_dict中
    val_image_dict[image_name] = image_id

    
# 创建一个新的空字典
new_val_image_dict = {}

# 遍历原字典的键值对
for key, value in val_image_dict.items():
    # 去掉键中的.jpg
    new_key = key.replace(".jpg", "")
    # 把新的键值对添加到新字典中
    new_val_image_dict[new_key] = value
new_val_image_dict

{'200003803_00002_2': 0,
 '200003803_00003_1': 1,
 '200003803_00003_2': 2,
 '200003803_00004_1': 3,
 '200003803_00004_2': 4,
 '200003803_00005_1': 5,
 '200003803_00006_1': 6,
 '200003803_00006_2': 7,
 '200003803_00007_1': 8,
 '200003803_00007_2': 9,
 '200003803_00008_1': 10,
 '200003803_00008_2': 11,
 '200003803_00009_1': 12,
 '200003803_00009_2': 13,
 '200003803_00010_1': 14,
 '200003803_00010_2': 15,
 '200003803_00011_2': 16,
 '200003803_00012_1': 17,
 '200003803_00012_2': 18,
 '200003803_00013_1': 19,
 '200003803_00013_2': 20,
 '200003803_00014_1': 21,
 '200003803_00014_2': 22,
 '200003803_00015_1': 23,
 '200003803_00015_2': 24,
 '200003803_00016_2': 25,
 '200003803_00017_1': 26,
 '200003803_00017_2': 27,
 '200003803_00018_1': 28,
 '200003803_00018_2': 29,
 '200003803_00019_1': 30,
 '200003803_00019_2': 31,
 '200003803_00020_1': 32,
 '200003803_00021_1': 33,
 '200003803_00021_2': 34,
 '200003803_00022_1': 35,
 '200003803_00022_2': 36,
 '200003803_00023_1': 37,
 '200003803_00023_2': 

In [43]:
len(new_val_image_dict)

50

### 生成annotation字段

In [44]:
# 创建一个空的列表来存储coco数据集格式的annotations字段
coco_annotations = []

# 遍历merged_df中的每一行
for index, row in val_df.iterrows():
    # 获取当前行的照片名，类别名，坐标和面积
    image_name = row["Image"]
    category_name = row["Unicode"]
    bbox = [row["X"], row["Y"], row["Width"], row["Height"]]
    area = row["Width"] * row["Height"]
    # 创建一个字典，存储当前行的信息，包括id，image_id，category_id，bbox和area
    
    annotation_info = {
        "id": index, # id从0开始递增
        "image_id": new_val_image_dict[image_name], # 根据照片名在new_image_dict中查找对应的id
        "category_id": category_dict[category_name], # 根据类别名在category_dict中查找对应的id
        "bbox": bbox,
        "area": area,
        "iscrowd": 0
    }
    # 将字典追加到coco_annotations列表中
    coco_annotations.append(annotation_info)

In [45]:
val_dict_to_save = {'annotations':coco_annotations, 'images':coco_val, 'categories':categories}

In [46]:
import json
# 日语字符以utf-8编码了，可以正常读取！
with open("val.json", 'w',encoding='utf-8') as f:
    json.dump(val_dict_to_save, f)

## 验证

In [47]:
# 导入COCO API
from pycocotools.coco import COCO

# 指定数据集和标注文件的路径
dataDir = "val"
annFile = "{}.json".format(dataDir)

# 创建COCO对象，加载标注文件
coco = COCO(annFile)

# 获取所有的图像id
imgIds = coco.getImgIds()

# 获取所有的类别id和名称
catIds = coco.getCatIds()
catNms = coco.loadCats(catIds)

# 打印一些信息
print("Number of images: {}".format(len(imgIds)))
print("Number of categories: {}".format(len(catIds)))
print("Category names: {}".format([cat["name"] for cat in catNms]))

loading annotations into memory...
Done (t=0.03s)
creating index...
index created!
Number of images: 50
Number of categories: 4328
Category names: ['U+0031', 'U+0032', 'U+0034', 'U+0036', 'U+0039', 'U+003F', 'U+2000B', 'U+20D45', 'U+2123D', 'U+22999', 'U+22CE3', 'U+231C3', 'U+2324A', 'U+233D0', 'U+23D0E', 'U+241C6', 'U+242CE', 'U+2437D', 'U+24E30', 'U+24E60', 'U+24FA3', 'U+2564A', 'U+25750', 'U+25877', 'U+25A0', 'U+25A1', 'U+25B2', 'U+25B3', 'U+25CB', 'U+25CF', 'U+25DA1', 'U+25E2E', 'U+25EF', 'U+25FFB', 'U+264B3', 'U+266B0', 'U+27752', 'U+28263', 'U+29780', 'U+29DDA', 'U+29E75', 'U+3002', 'U+3005', 'U+3006', 'U+3007', 'U+300C', 'U+300D', 'U+3031', 'U+3032', 'U+303B', 'U+303D', 'U+3042', 'U+3044', 'U+3046', 'U+3048', 'U+304A', 'U+304B', 'U+304C', 'U+304D', 'U+304E', 'U+304F', 'U+3050', 'U+3051', 'U+3052', 'U+3053', 'U+3054', 'U+3055', 'U+3056', 'U+3057', 'U+3058', 'U+3059', 'U+305A', 'U+305B', 'U+305C', 'U+305D', 'U+305E', 'U+305F', 'U+3060', 'U+3061', 'U+3062', 'U+3063', 'U+3064', 'U+3